lib/index.d.ts

```diff
@@ -0,0 +1,214 @@
+import webpack = require("webpack");
+
+type WebpackUserscriptOptions =
+  WPUSOptions |
+  HeaderFile |    // shorthand for WPUSOptions.headers
+  HeaderProvider  // shorthand for WPUSOptions.headers
+
+interface WPUSOptions {
+  headers: HeaderFile | HeaderProvider | HeaderObject
+
+  /**
+   * Output *.meta.js files or not.
+   * Defaults to true.
+   */
+  metajs?: boolean
+
+  /**
+   * Rename all .js files to .user.js files or not.
+   * Defaults to true.
+   */
+  renameExt?: boolean
+
+  /**
+   * Prettify the header or not.
+   * Defaults to true.
+   */
+  pretty?: boolean
+
+  /**
+   * Base URL for downloadURL.
+   * If not provided, it will be set to `updateBaseUrl` if `updateBaseUrl` is provided.
+   */
+  downloadBaseUrl?: string
+
+  /**
+   * Base URL for updateURL.
+   * If not provided, it will be set to `downloadBaseUrl` if `downloadBaseUrl` is provided.
+   */
+  updateBaseUrl?: string
+
+  /**
+   * Looks similar to `*.meta.js` but with additional `@require` meta field to include the main userscript.
+   * It can be useful if you set TamperMonkey not to cache external files.
+   */
+  proxyScript?: {
+    /**
+     * Filename template of the proxy script.
+     * Defaults to "[basename].proxy.user.js".
+     */
+    filename: string
+
+    /**
+     * Base URL of the dev server.
+     * Defaults to "http://localhost:8080/".
+     */
+    baseUrl: string
+
+    /**
+     * Enable proxy script generation or not.
+     * Default value depends on whether `process.env.WEBPACK_DEV_SERVER` is `"true"` or not.
+     */
+    enable: boolean | (() => boolean)
+  }
+
+  /**
+   * Defaults to false.
+   */
+  ssri?: boolean | {
+    /**
+     * URL filters.
+     * Each of them is actually testing against a string compound of the meta field and the URL.
+     * For example, if a header is provided as `{ require: "http://example.com/sth.js" }`,
+     * a string of "// @require http://example.com/sth.js" is tested with the provided filters.
+     */
+    include: string | RegExp | string[] | RegExp[]
+    exclude: string | RegExp | string[] | RegExp[]
+
+    /**
+     * @see https://github.com/npm/ssri#--integritystreamopts---integritystream
+     */
+    algorithms: ("sha256" | "sha384" | "sha512")[]
+    integrity: string
+    size: number
+  }
+}
+
+type HeaderFile = string
+
+type HeaderProvider = (data: DataObject) => HeaderObject
+
+type HeaderObject = Record<string, string | Array<string>> & {
```

Comment:
I defined the `HeaderObject` a `Record<string, string | Array<string>>` for flexibility across userscript engines like TamperMonkey or GeaseMonkey, maybe there are some engine-specific headers too.

Listing out common headers is a good idea anyway to help developers easier to define the header, but I think it can be written like this to keep the flexibility,

```ts
type HeaderObject = {
    name?: string
    namespace?: string

    // ... other fields

    [field: string]: string | string[]  // for any other field that is not listed above
}
```

Reply:
Got it, I thought `  & ` accomplished that.